[RDY] Add TabNew, TabNewEntered, TabClosed events

fmoralesc opened the pull request:
TabNew triggers when entering a new tab page, but not when entering an
already created one.

EDIT: I also sent this to vim_dev.

fmoralesc committed:
Add TabNew event

TabNew triggers when entering a new tab page, but not when entering an
already created one.

fmoralesc committed:
Add TabNewEntered

TabNewEntered is triggered after vim has entered a buffer in new tab.

traycerb commented:
Similarly, what about a `TabClose` variant?  I [asked about something like this on vim_dev](https://groups.google.com/forum/#!topic/vim_use/rEC1OjsP0Fs), but there wasn't any feedback.

fmoralesc commented:
@traycerb That seems like a good idea. I've been working on other stuff, but I was planning on going back to this later today. I think I'll split the tab related event from the window related ones. The second have been much more of a PITA.

fmoralesc committed:
Add TabClosed event

TabClosed is triggered when a tab page closes.

fmoralesc commented:
An update: I've fixed the implementation of the `Tab*` events, so now they really match on filenames, and added some tests. I haven't looked at the `Win*` events yet. 

I also implemented `TabClose`. I'm uncertain about what should be the proper value for `<afile>` there. Currently it is the name of the file in the current buffer of the tab closing, but perhaps the tab page number would be more useful? Also, the test for TabClose is currently incomplete, because without [patch 530](https://code.google.com/p/vim/source/detail?r=v7-4-530) I can't do `2tabonly` (foe example), which I would want to check that the tab numbers are updated correctly. I might make a separate PR for it.

Once I'm happy with this, I'm going to split the history so every commit touches on only one event. That way you can cherry-pick them in case some of these events are unwanted (if any is wanted at all :p).

fmoralesc commented:
I decided to reduce the scope of this to the Tab* events, I don't think I'll be able to advance much with the Win* events because of lack of time (also, the other PRs, like #1793 and #1667). I'm marking as RFC too because I think this set of events is behaving well.

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -859,6 +860,10 @@ TabEnter			Just after entering a tab page. |tab-page|
 				first.
+							*TabNew*
+TabNew                          Before entering a new tab page. |tab-page|
+                                After WinEnter and before TabEnter.
```
One thing that drives me crazy about the existing autocmd documentation is that it is not clear when "new" events happen in relation to "enter" events. So it would be great to start documenting this. Does `TabNew` happen *before* or *after* `TabEnter`?

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -859,6 +860,10 @@ TabEnter			Just after entering a tab page. |tab-page|
+							*TabNew*
+TabNew                          Before entering a new tab page. |tab-page|
+                                After WinEnter and before TabEnter.
+
```
Should use tabs for indentation in the docs, unfortunately. Otherwise there will be layout issues

justinmk replied:
> WinEnter and before TabEnter

So the order is: 

    WinEnter > TabNew > TabNewEntered > TabEnter

can you document that and add a test for it?

fmoralesc replied:
Sure ;) BTW, The documentation on this was updated for vanilla vim by Bram because of my findings here. This is included in patches #1793 .

justinmk commented on src/nvim/fileio.c:
```diff
@@ -5245,8 +5245,11 @@ static struct event_name {
   {"TabEnter",        EVENT_TABENTER},
   {"TabLeave",        EVENT_TABLEAVE},
+  {"TabNew",          EVENT_TABNEW},
+  {"TabNewEntered",   EVENT_TABNEWENTERED},
```
Shouldn't there be an entry in `autocmd.txt` for this?

fmoralesc replied:
:+1:

justinmk commented on test/functional/autocmd/tabnewentered_spec.lua:
```diff
@@ -0,0 +1,24 @@
+           end)
+        end)
+        describe('with FILE as <afile>', function()
+            it('matches when openinf a new tab for FILE', function()
```
"openinf" => "opening"

justinmk commented:
> Also, the test for TabClose is currently incomplete, because without patch 530 I can't do 2tabonly (foe example), which I would want to check that the tab numbers are updated correctly. I might make a separate PR for it.

Will wait for 530 to get merged.

justinmk commented:
One last nitpick: I see you named `TabNewEntered` with past tense, I think that's a good convention (and use present particle for "before", i.e. `TabNewEntering`). But then `TabClose` uses imperative voice.

I think either `TabClose` should be renamed to `TabClosed`, *or* `TabNewEntered` should be renamed to `TabNewEnter`. We can choose either approach since vim isn't consistent about any of it :/

fmoralesc commented:
I like `TabClosed` better, I think I used `TabClose` because of the original idea proposal by @traycerb.

fmoralesc commented:
OK. I updated the TabClosed tests to use the command ranges syntax from #1793, so it should fail until that is merged. In my local tests (rebasing over that set of patches) it behaves as expected.

fmoralesc commented:
Ah! It seems like I didn't need the full command ranges syntax, so this should be good to go as-is.

fmoralesc commented:
I sent the updated patches to vim_dev at https://groups.google.com/forum/#!topic/vim_dev/bduueOtT-4c

fmoralesc commented:
@justinmk Rebased, as requested in #1172

justinmk commented:
After reviewing `:help autocmd-events`, and after comparing all the different names, here are some thoughts. 

Only these events use the past tense:

```
EncodingChanged
TermChanged
GUIFailed
FileChangedShell
FileChangedShellPost
FileChangedRO
CmdUndefined
FuncUndefined
VimResized
FocusGained
CursorMoved
CursorMovedI
TextChanged
TextChangedI
```

`TextChanged` is fairly new, so it can't be said that the events are all legacy names.  Also, use of past tense doesn't necessary imply chronology because FileChangedShell and FileChangedShellPost both exist.

Only nitpick that can be made about this PR, I think, is that `TabNewEntered` is inconsistent with the other `FooEnter` events. If that bothers anyone, I wouldn't be opposed to changing it to `TabNewEnter`. But otherwise I think we have done a reasonable job here.

fmoralesc commented:
Sure. `TabNewEntered`'s name was bound to be a point of contention, since it works like a buffer event, but is more meaningful as a tab page event. `BufEnterOnNewTab` would be the more descriptive option (if only BufEnter wasn't triggered *also*), but it is a mouthful.

justinmk commented:
merged
